src/Microsoft.AspNetCore.Blazor.Server/Bots/RenderBatchWriter.cs

```diff
@@ -0,0 +1,235 @@
+
+            var edits = diff.Edits;
+            _binaryWriter.Write(edits.Count);
+
+            var editsArray = edits.Array;
+            var editsEndIndexExcl = edits.Offset + edits.Count;
+            for (var i = edits.Offset; i < editsEndIndexExcl; i++)
+            {
+                Write(editsArray[i]);
+            }
+        }
+
+        void Write(in RenderTreeEdit edit)
+        {
+            // We want all RenderTreeEdit outputs to be of the same length, so that
+            // the recipient can index into the array directly without walking it.
+            // So we output some value for all properties, even when not applicable
+            // for this specific RenderTreeEditType.
+            _binaryWriter.Write((int)edit.Type);
+            _binaryWriter.Write(edit.SiblingIndex);
+            _binaryWriter.Write(edit.ReferenceFrameIndex);
+            WriteString(edit.RemovedAttributeName);
+        }
+
+        int Write(in ArrayRange<RenderTreeFrame> frames)
+        {
+            var startPos = (int)_binaryWriter.BaseStream.Position;
+
+            var array = frames.Array;
+            var count = frames.Count;
+            _binaryWriter.Write(count);
+            for (var i = 0; i < count; i++)
+            {
+                Write(array[i]);
+            }
+
+            return startPos;
+        }
+
+        void Write(in RenderTreeFrame frame)
+        {
+            _binaryWriter.Write((int)frame.FrameType);
+
+            // We want each frame to take up the same number of bytes, so that the
+            // recipient can index into the array directly instead of having to
+            // walk through it.
+            // Since we can fit every frame type into 3 ints, use that as the
+            // common size. For smaller frames, we add padding to expand it to
+            // 12 bytes (i.e., 3 x 4-byte ints).
+            // The total size then for each frame is 16 bytes (frame type, then
+            // 3 other ints).
+            switch (frame.FrameType)
+            {
+                case RenderTreeFrameType.Attribute:
+                    WriteString(frame.AttributeName);
+                    WriteString(frame.AttributeValue as string);
+                    _binaryWriter.Write(frame.AttributeEventHandlerId);
+                    break;
+                case RenderTreeFrameType.Component:
+                    _binaryWriter.Write(frame.ComponentSubtreeLength);
+                    _binaryWriter.Write(frame.ComponentId);
+                    _binaryWriter.Seek(4, SeekOrigin.Current); // Padding
+                    break;
+                case RenderTreeFrameType.ComponentReferenceCapture:
+                    // The client doesn't need to know about these. But we still have
+                    // to include them in the array otherwise the ReferenceFrameIndex
+                    // values in the edits data would be wrong.
+                    _binaryWriter.Seek(12, SeekOrigin.Current); // Padding
```

Comment:
I think this might throw from non-seekable streams. Generally memory and file streams are seekable, and networking-related streams are not.

Reply:
Yes, it would throw in that case. Currently we only use this with `MemoryStream`, which is why it's currently OK.

Do you think it would be better to explicitly write zeros to the stream instead of seeking forwards like this?

Reply:
Writing zeros would be better since it would be portable across all stream types, and it doesn't seem too complicated